### ISSUE TITLE
Context Dependency Optimization and GPyOpt

### ISSUE THREAD

**Issue by AliBaheri** [SheffieldML/GPyOpt#29]
Hi,
In some optimization problem, some variable is not an optimization variable,
but something that is fixed by the environment.In the literature, this is
refereed to as context. One way to handle context is to define an 
additional input on kernel. Then, we fix the context and optimize the acquisition over the
remaining optimization variable. Does the GPyOpt can deal with the contexts? 
If not, can someone please give some hints to implement this idea? 
I want to implement something like this:

> bounds = [(bounds of opt_variable),(bounds of context)]
> k_parameters = GPy.kern.RBF(1, 1, 1, active_dims=[0])
> k_context = GPy.kern.RBF(1, 1, 1, active_dims=[1])
> kern = k_optimization_variable \* k_context
> GPyOpt.methods.BayesianOptimization(...,active_dims = [0])

In other words, how can I fix the context and optimize the acquisition only over optimization variable?

Thanks for your help in advance!

**Comment by AliBaheri** [comment]
Can someone please provide some hints for the above question? Thanks for your consideration.

**Comment by javiergonzalezh** [comment]
In the acquisition optimizer you can fix the values some of the (continuous) dimensions and optimize over the rest. You need to call (within the optimizer) ```.fix_dimensions(dims, values)`` where:
- dims is a list of indexes with the dimensions you want to fix.
- values: a list with the values at which you want to fix them.
If both, the dimensions and the values are know and fixed from the beginning, it all should be  matter of configuring the optimizer appropriately before running the optimization. 
Hope this help!

**Comment by AliBaheri** [comment]
Hi Javier!
Thanks for your comments.
I changed the "dims" to 1 in `.fix_dimensions(dims,values)`, but I steel say that this dimension has not been fixed.
Basically, I have one optimization variable and one context which should be fixed.
I have defined the my kernel in the following form:

```
k_parameters = GPy.kern.RBF(1,.1,.5, active_dims = [0])
k_context = GPy.kern.Linear(1, 1, 1, active_dims= [1])
kern = k_parameters * k_context
bounds = [(0.146,1.528),(0,240)]

myBopt = GPyOpt.methods.BayesianOptimization(f = None, bounds = bounds, kernel=kern,
X =np.hstack((x,context)), Y = fnvalue, acquisition_type='EI')
myBopt.run_optimization() 
a = myBopt.suggested_sample
```
using a for loop, at each iteration I am interested to know just the next best point by GPyOpt. Using this suggested sample, I can evaluate my black-box function, then agin pass to myBopt, and agin get the best next sample and so on...
However, at the first iteration I want that context to be fixed at 0, at second iteration to be fixed at 30, at third iteration to be fixed at 60 and so one and so forth...till let say 240.

Could you please give some hints how can I implement this with GPyOpt?

**Comment by AliBaheri** [comment]
So could you please clarify what do you mean by appropriately configuring the optimizer?
Do I need just change the "dims" and "values" in acquisition_optimizer.py? It did not work for some reason. 
Or I have to create the optimization object in the following form:
`aquisition_optimizer` = `GPyOpt.optimization.AcquisitionOptimizer()`

Plus, I need to fix the context (the fixed dimensions) at several points. At each of these points, I need to know the next suggested sample. Would you please give some hints in this regard?

Thanks for your help in advance!

**Comment by javiergonzalezh** [comment]
By changing ```dims``` and ```values``` in acquisition_optimizer.py you can fix the context but so far the code is written to use that option to deal with continuous variables as they were discrete (fixed to some specific values). You may need to write a some code to access the optimizer from the main wrapper class and fix your context. I can help you with that if you want.

**Comment by AliBaheri** [comment]
Thanks for your answer.
Would you please give some help to access the optimizer from the main wrapper class to fix the context?

**Comment by javiergonzalezh** [comment]
Hi Ali,

In the devel branch we have included a way of including a context variable in the optimization. It's still a bit preliminar but it may work for you. 

Javier
